parley/src/layout/cursor.rs

```diff
@@ -719,6 +734,42 @@ impl Selection {
         }
     }
 
+    /// Returns a new selection with the focus moved to just after the previous hard line break.
+    ///
+    /// If `extend` is `true` then the current anchor will be retained,
+    /// otherwise the new selection will be collapsed.
+    #[must_use]
+    pub fn hard_line_start<B: Brush>(&self, layout: &Layout<B>, extend: bool) -> Self {
+        if let Some((mut hard_line_start_index, line)) = self.focus.line(layout) {
+            let mut result_byte_index = line.text_range().start;
+            loop {
+                if hard_line_start_index == 0 {
+                    break;
+                }
+                let prev_index = hard_line_start_index - 1;
+                let Some(line) = layout.get(prev_index) else {
+                    unreachable!(
+                        "{hard_line_start_index} is a valid line in the layout, but {prev_index} isn't, despite the latter being smaller.\n\
+                        The layout has {} lines.",
+                        layout.len()
+                    );
```

Comment:
I don't think you need a message that detailed for an `unreachable!()` macro FWIW.